### PR TITLE
Fix task memory leak

### PR DESCRIPTION
**Pull request by thirstycrow**
Three issues that prevent completed tasks from destroy are fixed:
- The ref count was initialized to 1, incremented on schedule, and decremented after run, so it will never going to be 0
- ~~The ref count could be updated on a foreign executor, so the relax ordering is not enough~~
- There's a chance that a foreign waker is dropped when the task is already complete. An additional foreign wake is necessary in this case.

Related issue: #401

### CODE REVIEW

File: glommio/src/task/header.rs [hunk 1]
```diff
@@ -82,12 +89,38 @@ impl Header {
         // Put the waker into the awaiter field.
         abort_on_panic(|| self.awaiter = Some(waker.clone()));
     }
+
+    #[allow(dead_code)]
```
Comment (glommer) [inline glommio/src/task/header.rs:93]: isn't it better to make this conditional on `debugging`?

**Comment by glommer** [comment]
Why isn't relaxed enough? Being updated on a foreign executor should not require ordering stronger than relaxed.
Usually ordering is needed when there are other memory accesses that need to be synchronized before or after this address.
If all we're doing is testing the atomic variable, relaxed should be enough.

**Comment by thirstycrow** [comment]
You're right. I had some misunderstanding about the memory orderings.

File: glommio/src/task/header.rs (hunk 1 above)
Comment (thirstycrow) [reply glommio/src/task/header.rs:93]: Yes, that would be better.

**Review by glommer — APPROVED** [review]

**Comment by glommer** [comment]
@thirstycrow how slow is the task debugging thing you added? Any idea?

I'd love to see these tests executed as part of the `cargo test` default suite.

For now I am merging so we can fix the leak, but let's consider separately adding that to the main execution flow.

Thanks!

**Comment by thirstycrow** [comment]
> @thirstycrow how slow is the task debugging thing you added? Any idea?
It depends on how many tasks needs to be tracked. The default filter function let all labelled tasks pass. A custom one with lower sample rate could reduce the impact. In the benchmark for our app, requests latency raise from 2ms to 10s with the default filter function, and reduced to 10ms with a sample rate of 0.1

**Pull request merged**